#  Week 5 -Home work Solution

Go through this carefully (and run it yourself), as we will revist this problem in Week 7 with Ridge, Lasso & ELasticNet

In [38]:
#Import the packages

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from CBFV import composition

from sklearn.linear_model import LinearRegression

from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import normalize

In [14]:
data=pd.read_csv("data/lec05-HW_formation_energy.csv")
data.head()

,Unnamed: 0,jid,spg_number,spg_symbol,formula,formation_energy_peratom,func,optb88vdw_bandgap
0,1,JVASP-90856,129,P4/nmm,TiCuSiAs,-0.42762,OptB88vdW,0.000
1,2,JVASP-86097,221,Pm-3m,DyB6,-0.41596,OptB88vdW,0.000
2,3,JVASP-64906,119,I-4m2,Be2OsRu,0.04847,OptB88vdW,0.000
3,4,JVASP-98225,14,P2_1/c,KBi,-0.44140,OptB88vdW,0.472
4,5,JVASP-10,164,P-3m1,VSe2,-0.71026,OptB88vdW,0.000


In [15]:
len(data) #that's a lot of data

55716

In [16]:
data.isna().sum()

Unnamed: 0                  0
jid                         0
spg_number                  0
spg_symbol                  0
formula                     2
formation_energy_peratom    0
func                        0
optb88vdw_bandgap           0
dtype: int64

In [17]:
#There are two missing formulas, out of a total of 55K!, so let's just drop them

data=data.dropna(axis=0,how='any')
len(data)

55714

In [18]:
#There are a lot of un-necessary columns, let's get rid of some of them for convenience
data=data.drop(['Unnamed: 0','optb88vdw_bandgap','func','spg_symbol','spg_number','jid'],axis=1)
data.head()

,formula,formation_energy_peratom
0,TiCuSiAs,-0.42762
1,DyB6,-0.41596
2,Be2OsRu,0.04847
3,KBi,-0.44140
4,VSe2,-0.71026


In [21]:
#Remember: if we want to use CBFV to featurize the data, we need to have our columns as 'formula' and 'target'
data=data.rename(columns={'formation_energy_peratom':'target'})
data.head()

,formula,target
0,TiCuSiAs,-0.42762
1,DyB6,-0.41596
2,Be2OsRu,0.04847
3,KBi,-0.44140
4,VSe2,-0.71026


In [30]:
print(f'shape of data {data.shape}')

shape of data (55714, 2)


In [32]:
X,y,formulae,skipped=composition.generate_features(data,elem_prop='oliynyk')  
#You can try different elem_prop arguments such as -mat2vec, jarvis etc.

Processing Input Data: 100%|███████████████████████████████████████████████████| 55714/55714 [00:05<00:00, 9809.97it/s]


	Featurizing Compositions...


Assigning Features...: 100%|███████████████████████████████████████████████████| 55714/55714 [00:07<00:00, 7328.83it/s]



NOTE: Your data contains formula with exotic elements. These were skipped.
	Creating Pandas Objects...


In [33]:
X.head()


,avg_Atomic_Number,avg_Atomic_Weight,avg_Period,avg_group,avg_families,avg_Metal,avg_Nonmetal,avg_Metalliod,avg_Mendeleev_Number,avg_l_quantum_number,...,mode_polarizability(A^3),mode_Melting_point_(K),mode_Boiling_Point_(K),mode_Density_(g/mL),mode_specific_heat_(J/g_K)_,mode_heat_of_fusion_(kJ/mol)_,mode_heat_of_vaporization_(kJ/mol)_,mode_thermal_conductivity_(W/(m_K))_,mode_heat_atomization(kJ/mol),mode_Cohesive_energy
0,24.500000,53.608272,3.750000,11.000000,5.000000,0.500000,0.250000,0.250000,67.250000,1.500000,...,4.3,1090.15,890.15,2.33,0.33,13.050,34.76,21.90,302.0,2.960
1,13.714286,32.480857,2.571429,11.571429,5.571429,0.142857,0.000000,0.857143,66.142857,1.285714,...,3.0,2352.15,4275.00,2.34,1.02,50.200,489.70,27.00,573.0,5.810
2,32.000000,77.331090,3.750000,5.000000,3.000000,1.000000,0.000000,0.000000,61.750000,1.000000,...,5.6,1551.15,3243.15,1.85,1.82,7.950,292.40,200.00,324.0,3.320
3,51.000000,124.039335,5.000000,8.000000,3.000000,1.000000,0.000000,0.000000,44.500000,0.500000,...,7.4,336.40,1033.15,0.86,0.12,2.334,79.87,7.87,90.0,0.934
4,30.333333,69.620500,4.000000,12.333333,6.000000,0.333333,0.666667,0.000000,74.666667,1.333333,...,3.8,490.15,958.15,4.79,0.32,6.694,37.70,0.52,227.0,2.460


In [43]:
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.2)
len(X_train)

44571

In [37]:
scaler=StandardScaler()  #Instantiate a scaler

X_train_scaled=scaler.fit_transform(X_train)
X_test_scaled=scaler.transform(X_test)

In [40]:
model=LinearRegression()  # Create an instance of Linear Regression
result=model.fit(X_train,y_train)

In [41]:
r2=model.score(X_train,y_train)
print(r2)

0.8387758272337877


In [42]:
r2_test=model.score(X_test,y_test)
print(r2_test)

0.8421570701621841


## ‼️ 
This featurization exercise is similar to what we did in class. Just the dataset is different.

However, you might also notice that here, we didn't run into a problem that we had in-class.

When we fitted a linear regression model to the in-class data (Thermal_cond), the data overfitted significantly!
Remember? if not, go back and have a look [here](week05-3_inclass_problem.ipynb)

We said that the problem was the length of the dataset in comparison to the number of features
In-class problem: train data - 46 while no. of features - 132 (magpie featurizer)

The dataset $(n)$ is too small $<<$ compared to features $(p)$  
In that case the equations are underdetermined - you can always find a solution that fits best to the training data.
So we have to significantly drop featurizers so that $n>p$ for this to work.

Whereas, 

In this problem : train data - 44571, while no.of features - 264 (oliynik featurizer)
So, $n>>p$. 
Therefore we don't have the same problem - the fitting works


Keep this in mind, as we will use the same dataset in week 7, and sub-sample from the data to see how effectively you can use Ridge/Lasso & ElasticNet regression to manage overfitting through "Regularization".

Unlike Big data (such as those obtained from Social Media apps) often times, materials datasets are very small. In such cases these methods are useful to create meaningful machine learning models.

## ‼️ 